android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.java

```diff
@@ -0,0 +1,99 @@
+    }
+
+    ExperienceDBObject experienceDBObject = ExponentDB.experienceIdToExperienceSync(remoteMessage.getData().get("experienceId"));
+    if (experienceDBObject != null) {
+      try {
+        JSONObject manifest = new JSONObject(experienceDBObject.manifest);
+        int sdkVersion = ABIVersion.toNumber(manifest.getString("sdkVersion")) / 10000;
+
+        // If an experience is on SDK newer than 39, that is SDK40 and beyond up till UNVERSIONED
+        // we only use the new notifications API as it is going to be removed from SDK40.
+        if (sdkVersion >= 40) {
+          dispatchToNextNotificationModule(context, remoteMessage);
+          return;
+        } else if (sdkVersion == 38 || sdkVersion == 39) {
+          // In SDK38 and 39 we want to let people decide which notifications API to use,
+          // the next or the legacy one.
+          JSONObject androidSection = manifest.optJSONObject("android");
+          if (androidSection != null) {
+            boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(context, remoteMessage);
+              return;
+            }
+          }
+        }
+        // If it's an older experience or useNextNotificationsApi is set to false, let's use the legacy notifications API
+        dispatchToLegacyNotificationModule(context, remoteMessage);
+      } catch (JSONException e) {
+        e.printStackTrace();
+        EXL.e("expo-notifications", "Couldn't parse the manifest.");
+      }
+    } else {
+      EXL.e("expo-notifications", "No experience found for id " + remoteMessage.getData().get("experienceId"));
+    }
+  }
+
+  private void dispatchToNextNotificationModule(Context context, RemoteMessage remoteMessage) {
+    super.onMessageReceived(context, remoteMessage);
+  }
+
+  private void dispatchToLegacyNotificationModule(Context context, RemoteMessage remoteMessage) {
+    PushNotificationHelper.getInstance().onMessageReceived(context, remoteMessage.getData().get("experienceId"), remoteMessage.getData().get("channelId"), remoteMessage.getData().get("message"), remoteMessage.getData().get("body"), remoteMessage.getData().get("title"), remoteMessage.getData().get("categoryId"));
+  }
+
+  @NotNull
+  @Override
+  protected NotificationRequest createNotificationRequest(@NotNull String identifier, @NotNull NotificationContent content, FirebaseNotificationTrigger notificationTrigger) {
+    ExperienceId experienceId;
+    Map<String, String> data = notificationTrigger.getRemoteMessage().getData();
+    if (!data.containsKey("experienceId")) {
+      experienceId = null;
+    } else {
+      experienceId = ExperienceId.create(data.get("experienceId"));
+    }
+    String experienceIdString = experienceId == null ? null : experienceId.get();
+    return new ScopedNotificationRequest(identifier, content, notificationTrigger, experienceIdString);
+  }
+}
```

Comment:
Hmm, can't we go with kotlin here as well? 🤔 
I believe we're already mixing in the kotlin language into the `expoview`/`kernel` files 😁

Reply:
I wished to make review process easier for you by not refactoring this code to Kotlin in the same PR as this would introduce more non-essential changes (and what I did to create `ExpoFcmMessagingDelegate` was simply to change {file,class}name from `ExpoFcmMessagingService` and add `Context context` parameters). I, for one, would prefer to see two separate commits/PRs: one for copy-paste and another one for migration. Seeing that you have already reviewed this and this is the only concern I will migrate to Kotlin in a separate commit. 🙇